File: android/src/main/java/com/superwallreactnative/models/PaywallSkippedReason.kt
package com.superwallreactnative.models

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReadableMap
import com.superwall.sdk.paywall.presentation.internal.state.PaywallSkippedReason

class PaywallSkippedReason {
  companion object {
    fun toJson(skippedReason: PaywallSkippedReason): ReadableMap {
      val map = Arguments.createMap()

      when (skippedReason) {
        is PaywallSkippedReason.Holdout -> {
          map.putString("type", "Holdout")
          map.putMap("experiment", Experiment.toJson(skippedReason.experiment))
        }

        is PaywallSkippedReason.NoAudienceMatch -> {
          map.putString("type", "NoAudienceMatch")
        }

        is PaywallSkippedReason.PlacementNotFound -> {
          map.putString("type", "PlacementNotFound")
        }

        is PaywallSkippedReason.UserIsSubscribed -> {
          map.putString("type", "UserIsSubscribed")
        }
      }

      return map
    }
  }
}

File: ios/Json/PaywallSkippedReason.swift
//
//  PaywallSkippedReason.swift
//  Superwall
//
//  Created by Yusuf Tör on 21/02/2024.
//

import SuperwallKit

extension PaywallSkippedReason {
  func toJson() -> [String: Any] {
    switch self {
    case .holdout(let experiment):
      return [
        "type": "Holdout",
        "experiment": experiment.toJson()
      ]
    case .noAudienceMatch:
      return ["type": "NoAudienceMatch"]
    case .placementNotFound:
      return ["type": "PlacementNotFound"]
    }
  }
}
